Reproducible integration testing with simulated time

sergefdrv opened the pull request.

sergefdrv commented:
@xosmig @matejpavlovic I think you could already look into this draft PR. I just need to rebase and adjust it accordingly.

matejpavlovic reviewed (approved):
Great job!
It took me some effort to wrap my head around it, but it looks really good.
Only minor comments / questions.
Very well done, thanks!

sergefdrv commented on pkg/testsim/runtime.go:
```diff
@@ -0,0 +1,235 @@
+	r.now = e.deadline
+	r.queueLock.Unlock()
+	e.fn()
+	r.waitQuiescence()
```
Exactly, we need to wait for quiescence to complete the step because the invoked action can spawn new active processes.

matejpavlovic replied:
Right, got it. Would it make sense to not activate the processes directly, but schedule them at randomized instants in the future (virtual time - wise)?
E.g., if a process writes to several channels, it seems like the corresponding processes would always be activated in the same order (order of writing to the channels). If they were not directly activated, but only scheduled for future execution, could that capture more of the native non-determinism in Go's native scheduling?
This is just a thought about a potential extension. It definitely works the way it is and unless the extension makes sense and its implementation is trivial, let's leave it for later.

Pls. mark this comment as resolved when you read this.

sergefdrv replied:
I was thinking about that, but the difficulty is that the simulation still works around the Go runtime. With the current semantics, we cannot block in `Spawn` or `Fork`. Maybe we could redefine them such that they would work more like Go's `go` statement, i.e. receive a function as a parameter and run that function in the new process' context. However, I'm not sure how convenient that would be to use.

So far we explicitly `Yield` to avoid concurrency and `Delay` for random duration to simulate non-determinism of concurrent goroutines.

sergefdrv commented on pkg/testsim/process.go:
```diff
@@ -0,0 +1,123 @@
+
+// Process represents a context of execution in the simulation.
+type Process struct {
+	*Runtime
+	lock     sync.Mutex
```
Here, I actually prefer to use `Runtime` explicitly because it is not a part of `Process`, although it is conveniently declared as an embedded field, even exported for convenience of external packages. On the other hand, I don't want give external packages access to the lock, therefore made it an unexported field.

sergefdrv commented on pkg/deploytest/simulation.go:
```diff
@@ -0,0 +1,269 @@
+func (n *SimNode) Start(proc *testsim.Process, walEvents *events.EventList) {
+	n.SendEvents(proc, walEvents)
+	for m := range n.moduleChans {
+		initEvents := events.EmptyList()
+		initEvents.PushBack(events.Init(m))
+		n.SendEvents(proc, initEvents)
```
I decided to do the opposite change: moving `SendEvents` out of the loop.

sergefdrv commented on pkg/deploytest/simulation.go:
```diff
@@ -0,0 +1,269 @@
+	return sm
+}
+
+func (m *simModule) run(proc *testsim.Process, applyFn applyEventsFn) {
```
Good catch, I prefer to have less shared state, so removing the field.

sergefdrv commented on pkg/deploytest/simulation.go:
```diff
@@ -0,0 +1,269 @@
+		out.eventList, out.err = applyFn(in.ctx, in.eventList)
+
+		if out.err == nil {
+			it := origEvents.Iterator()
+			followUps := events.EmptyList()
+			for i := 0; i < in.eventList.Len(); i++ {
+				followUps.PushBackSlice(it.Next().Next)
+			}
+			origEvents = events.EmptyList()
+			for e := it.Next(); e != nil; e = it.Next() {
+				origEvents.PushBack(e)
+			}
+			followUps.PushBackList(out.eventList)
```
Honestly, I often find the event lists quite inconvenient to use, but I suppose there was some rational to introduce them instead of simply using slices. I believe the usage of the iterator here is correct. The first loop goes over as many leading items in `origEvents` as there are events passed to `ApplyEvents` (`in.eventList`). For each of those items, it collects the attached follow-up events in the `followUps` event list. The second loop continues iterating through the rest of `origEvents` and appends the remaining there events to a new empty event list reassigned to `origEvents`. If the event lists were slices, it would probably look like follows:

```go
var followUps []*eventpb.Event
for range in.eventList {
        followUps = append(followUps, origEvents[0].Next...)
        origEvents[0], origEvents = nil, origEvents[1:]
}
followUps = append(followUps, out.eventList...)
```

sergefdrv commented:
Summary of [changes](https://github.com/filecoin-project/mir/compare/003d7b16c6058023d777327c13dc3e118742c80a..09a513a2492eee9c57d7afba182377cff4afdf72): addressing review comments.
